### DISCUSSION TITLE
Next.js build: RangeError: Maximum call stack size exceeded when collecting build traces

### DISCUSSION THREAD

**Discussion by hongdeyuan** [vercel/next.js#58372] [Help]
### Summary

Why did the next build report an error when I did not use output: 'standalone'?

- not use output: 'standalone'
`yarn build`

`next version: 14.0.1 `

`next.config.js`
```
const withTM = require('next-transpile-modules')(['echarts', 'zrender']);

const nextConfig = {
compiler: {
    emotion: {
      sourceMap: true,
      autoLabel: 'never',
    },
  },
  images: {
    minimumCacheTTL: 3600,
  },
  experimental: {
    optimizePackageImports: [],
  },
  poweredByHeader: false,
}

module.exports = withTM(nextConfig)
```

Error screenshot：
![image](https://github.com/vercel/next.js/assets/37143703/b80ad1cc-0cff-430f-97cb-76291da2a580)

- use output: 'standalone'
build success 
![image](https://github.com/vercel/next.js/assets/37143703/2352ea93-5a57-4a68-b9f6-b5453aafe6cc)
![image](https://github.com/vercel/next.js/assets/37143703/c2bb5b91-c99e-4514-b21d-a7faf84b12bd)

### Additional information

```js
I don't know. Does anyone know why？

Is it because I didn't configure output？ [next config output](https://nextjs.org/docs/pages/api-reference/next-config-js/output)
```

**Comment by hongdeyuan** [comment]
It seems to be caused by `outputFileTracing` ？
Set `outputFileTracing: false` , build successfully;
😯

**Reply by dross15** [reply]
I am having the same issue with the maximum call stack, and while setting `output: 'standalone'` works, setting `outputFileTracing: false` does not build successfully. Static page generation times out. This only started happening moving from Next 13 -> 14.

**Reply by hongdeyuan** [reply]
What is your next.config.js configuration?

**Reply by hongdeyuan** [reply]
my next.config.js configuration：
```
{
  outputFileTracing: false, // setting here outputFileTracing
  swcMinify: true,
  compiler: {
    emotion: {
      sourceMap: true,
      autoLabel: 'never',
    },
  },
  images: {
    minimumCacheTTL: 3600,
  },
}
```

Additionally, you can try using [outputFileTracingExcludes](https://nextjs.org/docs/app/api-reference/next-config-js/output#caveats)? Perhaps your project has multiple sub modules? Try it? 
@dross15 

I found a relevant issue: https://github.com/vercel/next.js/issues/42641

**Reply by dross15** [reply]
@hongdeyuan setting `outputFileTracing: false` does work - but I'm not sure what this actually does (seems like when deploying with Docker it includes all files from your package's dependencies)? I'm curious if it may be breaking the app at all?

**Reply by hongdeyuan** [reply]
you're right. This feature(automatically trace) helps reduce the size of deployments drastically. Previously, when deploying with Docker you would need to have all files from your package's dependencies installed to run next start;

For details, please refer to [nextjs output](https://nextjs.org/docs/pages/api-reference/next-config-js/output)

**Reply by hongdeyuan** [reply]
However, I am also curious why `outputFileTracing: true` will fail to build; I haven't found the root cause yet；🤔

**Reply by dross15** [reply]
@hongdeyuan just curious, were you able to figure out anything more?

**Reply by Kolahzary** [reply]
This issue is also blocking us from upgrading next.js 13 to next.js 14.

We're using NX monorepo for all of our projects.

**Reply by hongdeyuan** [reply]
@dross15 
Recently, it was found that `@sentry/nextjs` did not upgrade from 6. x to 7. x, which caused this error;
fix:
![image](https://github.com/vercel/next.js/assets/37143703/fee30dcc-b110-49b9-aff8-9e0479cfe7c3)
before：
![image](https://github.com/vercel/next.js/assets/37143703/9cb3b1f6-aa6f-4bf4-875e-c28eb1c54804)
now:
![image](https://github.com/vercel/next.js/assets/37143703/d4f97152-eb9a-4e14-a85c-605135ce797c)

**Reply by dross15** [reply]
@hongdeyuan thanks for following up! Unfortunately I don't use `sentry` but perhaps it could be another dependency that's not up to date.

**Reply by aymendev1** [reply]
@hongdeyuan  Got the same issue and the `outputFileTracing: false` fixed it in NextJS 14.0.4 , but it seems that it might be an issue in the future due to the build message `Disabling outputFileTracing will not be an option in the next major version`

**Reply by veera20k** [reply]
anybody fixed it without using outputFileTracing: false?

**Reply by crudh** [reply]
@veera20k no, that is the only way I have found so far. I even generated a new fresh Next app and tried deploying it and got the same error. Turning that off fixes it.

**Reply by veera20k** [reply]
just now: I did downgrade the Sharp npm from 0.33.2 to 0.33.1, and it is working.

**Reply by soapproject** [reply]
@veera20k you are my hero!

**Reply by crudh** [reply]
@veera20k thanks! worked good for us too 👍

**Reply by titouancreach** [reply]
for us too, thank you so much !!

**Reply by krnrnee** [reply]
Down grading sharp from 0.33.2 to 0.33.1 worked for me as well. For additional info, the build worked fine locally, just not on Vercel until I downgraded sharp.

**Reply by dmarcucci** [reply]
Same deal for me with `sharp` going from `0.33.2` back down to `0.33.1`.

**Reply by mhughdo** [reply]
@veera20k Thank you so much. I got the same error, didn't know it was due to the new version of Sharp.

**Reply by krnrnee** [reply]
For clarification, I tried it on another website and it did build, though there were a bunch of sharp related messages pertaining to invalid linux settings or something.  I think what is happening is that Sharp is creating a ton of messages and if you have a lot of other stuff going on, it pushes you over the limit.  Regardless, downgrading fixed it for me on 2 separate installations.

**Reply by sattwyk** [reply]
Oh my goodness, thank you so much! You’re a lifesaver. 🙏

**Reply by ell-ska** [reply]
what version of `next` are you guys on? i'm on version `^14.0.4`. i tried to downgrade `sharp` to `0.33.1` and am still getting the same error

**Reply by sattwyk** [reply]
I’m on `next 14.0.4` and `sharp 0.33.1`. My Next.js app builds successfully now, but I’m getting an internal server error because the Sharp module for Linux-x64 is not being loaded.

```
Error: Could not load the "sharp" module using the linux-x64 runtime
Possible solutions:
- Ensure optional dependencies can be installed:
    npm install --include=optional sharp
    yarn add sharp --ignore-engines
- Ensure your package manager supports multi-platform installation:
    See https://sharp.pixelplumbing.com/install#cross-platform
- Add platform-specific dependencies:
    npm install --os=linux --cpu=x64 sharp
    npm install --force @img/sharp-linux-x64
- Consult the installation documentation:
    See https://sharp.pixelplumbing.com/install
    at Object.<anonymous> (/var/task/node_modules/.pnpm/sharp@0.33.1/node_modules/sharp/lib/sharp.js:107:9)
    at Module._compile (node:internal/modules/cjs/loader:1256:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1310:10)
    at Module.load (node:internal/modules/cjs/loader:1119:32)
    at Module._load (node:internal/modules/cjs/loader:960:12)
    at r.<computed>.e._load (/opt/node-bridge/bridge-server-RYVKFVW7.js:1:1574)
    at Module.require (node:internal/modules/cjs/loader:1143:19)
    at b.require (/var/task/node_modules/.pnpm/next@14.0.4_react-dom@18.2.0_react@18.2.0/node_modules/next/dist/compiled/next-server/server.runtime.prod.js:11:25829)
    at require (node:internal/modules/cjs/helpers:119:18)
    at Object.<anonymous> (/var/task/node_modules/.pnpm/sharp@0.33.1/node_modules/sharp/lib/constructor.js:10:1)
```
So I tried adding this to my `package.json` because I'm using `pnpm` so it can download the correct `sharp` module for different environments.

```
 "pnpm": {
    "supportedArchitectures": {
      "os": [
        "win32",
        "darwin",
        "linux",
        "current"
      ],
      "cpu": [
        "x64",
        "arm64"
      ]
    }
  }
  ```
  But still getting the internal server error.

**Reply by ell-ska** [reply]
actually i downgraded to `^0.33.1` and not `0.33.1`. after switching to `0.33.1` it works! here's a [discussion at vercel](https://github.com/orgs/vercel/discussions/5449) about how they are handling the issue

**Reply by sattwyk** [reply]
I have the same version `0.33.1` not `^0.33.1` and it builds successfully but now it's giving me a different error in function logs for not loading platform specific sharp module. I've created a issue here https://github.com/vercel/vercel/issues/11052#issue-2084734166. What package manager are you using tho ?

**Reply by dross15** [reply]
Are there any other culprits besides `sharp`? I am not using `sharp` and am still facing this issue.

**Reply by krnrnee** [reply]
Per Vercel, [5449](https://github.com/orgs/vercel/discussions/5449) upgrading to Next 14.1.0 fixes this issue and you can use the latest version of Sharp (0.33.2).  It worked for me.

**Reply by maxy4u** [reply]
A common cause of this error is rendering a server component as a child of a client component.

I accidentally  used "use client " in one component and there were server components present as its child. It took hours for me to identify this root cause :( 
Hope this help someone :)

**Comment by dariusrosendahl** [comment]
Also confirming that dowgrading `sharp` to `0.33.1` works!

**Comment by maycomayco** [comment]
Just another confirmation that **sharp 0.33.1** works well! 🫡

**Comment by beyera** [comment]
Echoing the chorus: `sharp` needed a throwback to `0.33.1`.

**Comment by bernabedev** [comment]
X2 Just another confirmation that sharp 0.33.1 works well! 👍

**Comment by LuciKritZ** [comment]
I removed ```sharp``` and it started working fine. Besides, I am using Sanity.io so I am still not sure if packages ```next-sanity``` or ```sanity``` are creating issues for ```sharp```

**Reply by fabohax** [reply]
did same and workd

**Comment by jupiterhub** [comment]
Upgrading to Next `14.1.0` works with sharp `0.33.2`

see: https://github.com/orgs/vercel/discussions/5449#discussioncomment-8199126

**Reply by Orpheus29** [reply]
Thank you so much. Finally, it works. Sharp 0.33.2 didn't want to collaborate with Next 14..0.3, producing this uninformative mistake. Bad sharpie!

![image](https://github.com/vercel/next.js/assets/110335429/ad12cb82-661e-4af9-9212-21d25931556d)

**Comment by maxy4u** [comment]
> Are there any other culprits besides `sharp`? I am not using `sharp` and am still facing this issue.
A common cause of this error is rendering a server component as a child of a client component.

I accidentally used "use client " in one component and there were server components present as its child. It took hours for me to identify this root cause :(
Hope this help someone :)

**Comment by liamrathke** [comment]
TL;DR: Remove Sharp unless you're sure you need it, even if the errors aren't Sharp related at all.

I ran into a similar issue, but with slightly different error messages when running `pnpm build`:

```
test-app:build: RangeError: Maximum call stack size exceeded
test-app:build:
test-app:build: RangeError: Maximum call stack size exceeded
test-app:build:     at walk (test-monorepo/node_modules/next/dist/compiled/@vercel/nft/index.js:1:59747)
test-app:build:     at Object.BinaryExpression (test-monorepo/node_modules/next/dist/compiled/@vercel/nft/index.js:1:60591)
test-app:build:     at walk (test-monorepo/node_modules/next/dist/compiled/@vercel/nft/index.js:1:59784)
test-app:build:     at Object.BinaryExpression (test-monorepo/node_modules/next/dist/compiled/@vercel/nft/index.js:1:60591)
test-app:build:     at walk (test-monorepo/node_modules/next/dist/compiled/@vercel/nft/index.js:1:59784)
test-app:build:     at Object.BinaryExpression (test-monorepo/node_modules/next/dist/compiled/@vercel/nft/index.js:1:60591)
test-app:build:     at walk (test-monorepo/node_modules/next/dist/compiled/@vercel/nft/index.js:1:59784)
test-app:build:     at Object.BinaryExpression (test-monorepo/node_modules/next/dist/compiled/@vercel/nft/index.js:1:60591)
test-app:build:     at walk (test-monorepo/node_modules/next/dist/compiled/@vercel/nft/index.js:1:59784)
test-app:build:     at Object.BinaryExpression (test-monorepo/node_modules/next/dist/compiled/@vercel/nft/index.js:1:60591)
test-app:build:  ELIFECYCLE  Command failed with exit code 1.
test-app:build: ERROR: command finished with error: command (test-monorepo/apps/test-app) /opt/homebrew/bin/pnpm run build exited (1)
test-app#build: command (test-monorepo/apps/test-app) /opt/homebrew/bin/pnpm run build exited (1)

 Tasks:    0 successful, 1 total
Cached:    0 cached, 1 total
  Time:    1m51.67s
Failed:    test-app#build
```

We were only getting the `RangeError` when building locally on macOS; our CI/CD builds didn't throw any errors, and building in Docker/Linux seemed to work as well. This was in a monorepo, and adjacent apps with basically identical configurations built fine as well across all environments. 

After trying to adjust the Node environment with things like `--stack_size=4000` (didn't work) and [adjusting the Sharp version](https://github.com/orgs/vercel/discussions/5449#discussioncomment-8150734) (didn't work), I realized that we weren't actually using Sharp at all, so I just removed it from the `package.json` and somehow everything built normally again. 

<img width="241" alt="Screenshot 2024-05-07 at 17 55 28" src="https://github.com/vercel/next.js/assets/33555592/2ed35f59-d338-4ba5-898a-4bafb7f283a7">
